chain/watcher.go

```diff
@@ -108,27 +110,12 @@ func (c *Watcher) maybeIndexTipSet(ctx context.Context, ts *types.TipSet) error
 	select {
 	case <-ctx.Done():
 		return ctx.Err()
-	case c.indexSlot <- struct{}{}:
-		// Indexing slot was available which means we can continue.
-		go func() {
-			// Clear the slot when we have completed indexing
-			defer func() {
-				<-c.indexSlot
-			}()
-
+	default:
+		c.wp.Submit(func() {
 			if err := c.obs.TipSet(ctx, ts); err != nil {
 				log.Errorw("failed to index tipset", "error", err, "height", ts.Height())
 			}
-		}()
-	default:
-		// The indexer is taking longer than one epoch to process. We need to avoid blocking the stream of incoming
-		// tipsets otherwise we will cause the node to fall behind the chain while it waits for us to catch up
-		// (which may never happen if we consistently take too long)
-		log.Errorw("skipping tipset since indexer is not ready", "height", ts.Height())
-		stats.Record(ctx, metrics.TipSetSkip.M(1))
-		if err := c.obs.SkipTipSet(ctx, ts, "indexer not ready"); err != nil {
```

Comment:
Calling SkipTipSet from outside the indexer probably isn't a good idea since it can cause off by one error as outlined in #705. Skip could also set `lastTipSet` when called.